# Patel Smitkumar Kaushikkumar                        (CIN: 306587208)

# Question1: Predicting Heart Disease

## a- Read the dataset file “Hearts_s.csv” (you should download it from CSNS), and assign it to a Pandas DataFrame.

In [73]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier

from sklearn.model_selection import train_test_split

from sklearn.metrics import accuracy_score

import numpy as np
import pandas as pd

heart_data = pd.read_csv('S:\MASTERS\SEMESTER_4\CS_4661\HomeWork\Homework_3\Heart_s.csv')
heart_data[0::10]

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
10    57      f  asymptomatic     140   192        0    148      0.4   
20    64      f       typical     110   211        2    144      1.8   
30    69      m       typical     140   239        0    151      1.8   
40    65      m  asymptomatic     150   225        2    114      1.0   
50    41      m    nontypical     105   198        0    168      0.0   
60    51      m  asymptomatic     130   305        0    142      1.2   
70    65      m    nonanginal     155   269        0    148      0.8   
80    45      f  asymptomatic     104   208        2    148      3.0   
90    62      m  asymptomatic     160   164        2    145      6.2   
100   34      f       typical     118   182        2    174      0.0   
110   56      f  asymptomatic     125   249        2    144      1.2   
120   63      m  asymptomatic     150   407        2    154      4.0   
130   51      f    nonanginal      94   227        0    154      0.0   
140   59      f       typical     170   288        2    159      0.2   
150   42      m  asymptomatic     102   265        2    122      0.6   
160   77      f  asymptomatic     125   304        2    162      0.0   
170   53      f  asymptomatic     142   226        2    111      0.0   
180   56      m  asymptomatic     134   409        2    150      1.9   
190   51      f  asymptomatic     140   298        0    122      4.2   
200   64      m  asymptomatic     180   325        0    154      0.0   
210   38      f       typical     120   231        0    182      3.8   
220   54      m    nonanginal     108   267        2    167      0.0   
230   55      m  asymptomatic     180   327        1    117      3.4   
240   41      m    nontypical     126   306        0    163      0.0   
250   58      f  asymptomatic     146   218        0    105      2.0   
260   58      m    nontypical     136   319        2    152      0.0   
270   46      f  asymptomatic     140   311        0    120      1.8   
280   55      m  asymptomatic     128   205        1    130      2.0   
290   44      f  asymptomatic     120   169        0    144      2.8   
300   38      f    nonanginal     138   175        0    173      0.0   

           Thal  AHD  
0         fixed   No  
10        fixed   No  
20       normal   No  
30       normal   No  
40   reversable  Yes  
50       normal   No  
60   reversable  Yes  
70       normal   No  
80       normal   No  
90   reversable  Yes  
100      normal   No  
110      normal  Yes  
120  reversable  Yes  
130  reversable   No  
140  reversable  Yes  
150      normal   No  
160      normal  Yes  
170  reversable   No  
180  reversable  Yes  
190  reversable  Yes  
200      normal   No  
210  reversable  Yes  
220      normal   No  
230      normal  Yes  
240      normal   No  
250  reversable  Yes  
260      normal  Yes  
270  reversable  Yes  
280  reversable  Yes  
290       fixed  Yes  
300      normal   No

## b - Check out the dataset. As you see, the dataset contains a number of features including both contextual and biological factors(e.g. age, gender, vital signs, …). The last column“AHD” is the label with “Yes” meaning that a human subject has Heart  Disease, and “No” meaning that the subject does not have Heart Disease.

In [74]:
def categorical_to_numerical(x):
    if x == 'Yes':
        return 1
    elif x == 'No':
        return 0
heart_data['label'] = heart_data['AHD'].apply(categorical_to_numerical)

y = heart_data['label']
y.head()

0    0
1    1
2    1
3    0
4    0
Name: label, dtype: int64

## c- As you see, there are at least 3 categorical features in the dataset (Gender, ChestPain,Thal). Let’s ignore these categorical features for now, only keep the numerical features and build your feature matrix and label vector.

In [75]:
feature_cols = ['Age', 'RestBP', 'Chol', 'RestECG', 'MaxHR', 'Oldpeak']
    
X = heart_data[feature_cols] 
X.head()

Age  RestBP  Chol  RestECG  MaxHR  Oldpeak
0   63     145   233        2    150      2.3
1   67     160   286        2    108      1.5
2   67     120   229        2    129      2.6
3   37     130   250        0    187      3.5
4   41     130   204        2    172      1.4

## d- Split the dataset into testing and training sets with the following parameters: test_size=0.25, random_state=4.

In [76]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=4)

## e- Use KNN (with k=3), Decision Tree (with random_state=5), and Logistic Regression Classifiers to predict Heart Disease based on the training/testing datasets that you built in part (d). Then check, compare, and report the accuracy of these 3 classifiers. Which one is the best? Which one is the worst?

In [77]:
k=3
knn = KNeighborsClassifier(n_neighbors = k)
knn.fit(X_train, y_train)
y_predict_knn = knn.predict(X_test)
accuracy_score_knn = accuracy_score(y_test, y_predict_knn)
print('accuracy for KNN: ', accuracy_score_knn )

logreg = LogisticRegression()
logreg.fit(X_train, y_train)
y_predict_lr = logreg.predict(X_test)
accuracy_score_lr = accuracy_score(y_test, y_predict_lr)
print('accuracy for Logistic Regression: ', accuracy_score_lr)

decisiontree = DecisionTreeClassifier(random_state=5)
decisiontree.fit(X_train, y_train)
y_predict_dt = decisiontree.predict(X_test)
accuracy_score_dt = accuracy_score(y_test, y_predict_dt)
print('accuracy for Decisiontree: ', accuracy_score_dt)

accuracy for KNN:  0.6973684210526315
accuracy for Logistic Regression:  0.7236842105263158
accuracy for Decisiontree:  0.6052631578947368


### BEST CLASSIFIER IS  : Logistic Regression with accuracy = 0.7236842105263158

### WORST CLSSIFIER IS  : Decisiontree  with accuracy = 0.6052631578947368

## f-Now, we want to use the categorical features as well! To this end, we have to perform a feature engineering process called OneHotEncoding for the categorical features. To do this, each categorical feature should be replaced with dummy columns in the feature table (one column for each possible value of a categorical feature), and then encode it in a binary manner such that only one of the dummy columns can take “1” at a time (and zero for the rest). For example, “Gender” can take two values “m” and “f”. Thus, we need to replace this feature (in the feature table) by 2 columns titled “m” and “f”. Wherever we have a male subject, we can put “1” and ”0” in the columns “m” and “f”. Wherever we have a female subject, we can put “0” and ”1” in the columns “m” and “f”. (Hint: you will need 4 columns to encode “ChestPain” and 3 columns to encode “Thal”).

In [78]:
# OneHotEncoding 
heart_data['male'] = heart_data['Gender'].map(lambda x: 1 if x=='m' else 0)
heart_data['female'] = heart_data['Gender'].map(lambda x: 1 if x=='f' else 0)

heart_data['typical'] = heart_data['ChestPain'].map(lambda x: 1 if x=='typical' else 0)
heart_data['asymptomatic'] = heart_data['ChestPain'].map(lambda x: 1 if x=='asymptomatic' else 0)
heart_data['nontypical'] = heart_data['ChestPain'].map(lambda x: 1 if x=='nontypical' else 0)
heart_data['nonanginal'] = heart_data['ChestPain'].map(lambda x: 1 if x=='nonanginal' else 0)

heart_data['fixed'] = heart_data['Thal'].map(lambda x: 1 if x=='fixed' else 0)
heart_data['normal'] = heart_data['Thal'].map(lambda x: 1 if x=='normal' else 0)
heart_data['reversable'] = heart_data['Thal'].map(lambda x: 1 if x=='reversable' else 0)

heart_data[0::10]

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
10    57      f  asymptomatic     140   192        0    148      0.4   
20    64      f       typical     110   211        2    144      1.8   
30    69      m       typical     140   239        0    151      1.8   
40    65      m  asymptomatic     150   225        2    114      1.0   
50    41      m    nontypical     105   198        0    168      0.0   
60    51      m  asymptomatic     130   305        0    142      1.2   
70    65      m    nonanginal     155   269        0    148      0.8   
80    45      f  asymptomatic     104   208        2    148      3.0   
90    62      m  asymptomatic     160   164        2    145      6.2   
100   34      f       typical     118   182        2    174      0.0   
110   56      f  asymptomatic     125   249        2    144      1.2   
120   63      m  asymptomatic     150   407        2    154      4.0   
130   51      f    nonanginal      94   227        0    154      0.0   
140   59      f       typical     170   288        2    159      0.2   
150   42      m  asymptomatic     102   265        2    122      0.6   
160   77      f  asymptomatic     125   304        2    162      0.0   
170   53      f  asymptomatic     142   226        2    111      0.0   
180   56      m  asymptomatic     134   409        2    150      1.9   
190   51      f  asymptomatic     140   298        0    122      4.2   
200   64      m  asymptomatic     180   325        0    154      0.0   
210   38      f       typical     120   231        0    182      3.8   
220   54      m    nonanginal     108   267        2    167      0.0   
230   55      m  asymptomatic     180   327        1    117      3.4   
240   41      m    nontypical     126   306        0    163      0.0   
250   58      f  asymptomatic     146   218        0    105      2.0   
260   58      m    nontypical     136   319        2    152      0.0   
270   46      f  asymptomatic     140   311        0    120      1.8   
280   55      m  asymptomatic     128   205        1    130      2.0   
290   44      f  asymptomatic     120   169        0    144      2.8   
300   38      f    nonanginal     138   175        0    173      0.0   

           Thal  AHD  label  male  female  typical  asymptomatic  nontypical  \
0         fixed   No      0     0       1        1             0           0   
10        fixed   No      0     0       1        0             1           0   
20       normal   No      0     0       1        1             0           0   
30       normal   No      0     1       0        1             0           0   
40   reversable  Yes      1     1       0        0             1           0   
50       normal   No      0     1       0        0             0           1   
60   reversable  Yes      1     1       0        0             1           0   
70       normal   No      0     1       0        0             0           0   
80       normal   No      0     0       1        0             1           0   
90   reversable  Yes      1     1       0        0             1           0   
100      normal   No      0     0       1        1             0           0   
110      normal  Yes      1     0       1        0             1           0   
120  reversable  Yes      1     1       0        0             1           0   
130  reversable   No      0     0       1        0             0           0   
140  reversable  Yes      1     0       1        1             0           0   
150      normal   No      0     1       0        0             1           0   
160      normal  Yes      1     0       1        0             1           0   
170  reversable   No      0     0       1        0             1           0   
180  reversable  Yes      1     1       0        0             1           0   
190  reversable  Yes      1     0       1        0             1           0   
200      normal   No

## g-Repeat parts (d) and (e) with the new dataset that you built in part (f). How does the prediction accuracy change for each method?

In [79]:
new_feature_cols = ['Age','RestBP','Chol','RestECG','MaxHR','Oldpeak', 'male', 'female', 'typical', 'asymptomatic',
                'nontypical', 'nonanginal','fixed', 'normal', 'reversable']

X1 = heart_data[new_feature_cols]
X_train, X_test,y_train,y_test = train_test_split(X1, y, test_size=0.25, random_state=4)

k=3
knn = KNeighborsClassifier(n_neighbors = k)
knn.fit(X_train, y_train)
y_predict_knn = knn.predict(X_test)
score_knn = accuracy_score(y_test, y_predict_knn)
print('accuracy for KNN:', score_knn)

logreg = LogisticRegression()
logreg.fit(X_train, y_train)
y_predict_lr = logreg.predict(X_test)
score_lr = accuracy_score(y_test, y_predict_lr)
print('accuracy for Logistic Regression:', score_lr)

decisiontree = DecisionTreeClassifier(random_state=5)
decisiontree.fit(X_train, y_train)
y_predict_dt = decisiontree.predict(X_test)
score_dt = accuracy_score(y_test, y_predict_dt)
print('accuracy for Decisiontree: ', score_dt)

accuracy for KNN: 0.6973684210526315
accuracy for Logistic Regression: 0.8026315789473685
accuracy for Decisiontree:  0.6973684210526315


## h- Now, repeat part (e) with the new dataset that you built in part (f), but this time using Cross-Validation. Thus, rather than splitting the dataset into testing and training, use 10-fold Cross-Validation (as we learned in Lab4) to evaluate the classification methods and report the final prediction accuracy.

In [80]:
from sklearn.cross_validation import cross_val_score

accuracy_list = cross_val_score(knn, X1, y, cv=10, scoring='accuracy')
print(accuracy_list)
accuracy_cv = accuracy_list.mean()
print('\n')
print('accuracy for Logistic KNN: ', accuracy_cv)

print("-------------------------------------------------------------------------------------------------------------")
accuracy_list = cross_val_score(logreg, X1, y, cv=10, scoring='accuracy')
print(accuracy_list)
accuracy_cv = accuracy_list.mean()
print('\n')
print('accuracy for Logistic Regression: ', accuracy_cv)

print("-------------------------------------------------------------------------------------------------------------")
accuracy_list = cross_val_score(decisiontree, X1, y, cv=10, scoring='accuracy')
print(accuracy_list)
accuracy_cv = accuracy_list.mean()
print('\n')
print('accuracy for Decisiontree: ', accuracy_cv)

[0.70967742 0.64516129 0.51612903 0.66666667 0.63333333 0.46666667
 0.66666667 0.66666667 0.55172414 0.72413793]


accuracy for Logistic KNN:  0.6246829810901001
-------------------------------------------------------------------------------------------------------------
[0.77419355 0.80645161 0.83870968 0.93333333 0.9        0.66666667
 0.8        0.83333333 0.79310345 0.79310345]


accuracy for Logistic Regression:  0.8138895068594735
-------------------------------------------------------------------------------------------------------------
[0.77419355 0.77419355 0.77419355 0.8        0.8        0.63333333
 0.63333333 0.53333333 0.65517241 0.82758621]


accuracy for Decisiontree:  0.7205339265850945
